transport/socks5/stream_dialer.go

```diff
@@ -30,23 +34,25 @@ type credentials struct {
 	password []byte
 }
 
-// NewStreamDialer creates a [transport.StreamDialer] that routes connections to a SOCKS5
+// NewDialer creates a [transport.StreamDialer] that routes connections to a SOCKS5
 // proxy listening at the given [transport.StreamEndpoint].
-func NewStreamDialer(endpoint transport.StreamEndpoint) (*StreamDialer, error) {
-	if endpoint == nil {
+func NewDialer(streamEndpoint transport.StreamEndpoint) (*Dialer, error) {
```

Comment:
That's because it seems SOCKS5 provides a ListenPacket interface, not a DialPacket interface. This is the same as Shadowsocks.
So the code should align with that, and we can create a dialer from a listener, just like we do with Shaodwsocks.

Reply:
@fortuna I have been digging into this; Yes, I believe that is the way to go too. Regarding the address included in the association request, it seems like that it must be either "0.0.0.0:0" for the unbound case or the address the client is listening on for the bound case (not the remote destination address). I can confirm this via the Socks server code lines below:

https://github.com/things-go/go-socks5/blob/d67aeab3543eddb5ea51a821ba6879e89f981d3e/handle.go#L222

Also, in the following test a listener is setup for the client and the client listening address it included in the associate request:

https://github.com/things-go/go-socks5/blob/d67aeab3543eddb5ea51a821ba6879e89f981d3e/server_test.go#L173

I am going to define `NewPacketListener, and `PacketListen` similar to Shadowsocks.

Reply:
@fortuna also looking at the `packet.go` and implementation of `boundPacketConn` in the SDK, It seems like the bound definition in the SDK is based on remote address binding where as in SOCK5 spec, binding refers to local address (client address) binding, i.e. the server only accepts packets from bound origin IP/Port arriving at the proxy side listening socket. I guess the bound connection concept can be applicable to both server and client side sockets and is a rather broad concept. 

```go
// Read implements [net.Conn].Read.
func (c *boundPacketConn) Read(packet []byte) (int, error) {
	for {
		n, remoteAddr, err := c.PacketConn.ReadFrom(packet)
		if err != nil {
			return n, err
		}
		if remoteAddr.String() != c.remoteAddr.String() {
			continue
		}
		return n, nil
	}
}
```